File: java/203.remove-linked-list-elements.java
/*
 * @lc app=leetcode id=203 lang=java
 *
 * [203] Remove Linked List Elements
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if (head == null) return null;
        ListNode be4Head = new ListNode(0), prev = be4Head;
        be4Head.next = head;
        while (head != null) {
            if (head.val == val) prev.next = head.next; 
            else prev = head;
            head = head.next;
        }
        return be4Head.next;
    }
}

File: go/203.remove-linked-list-elements.go
/*
 * @lc app=leetcode id=203 lang=golang
 *
 * [203] Remove Linked List Elements
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeElements(head *ListNode, val int) *ListNode {
	if head == nil {
		return nil
	}
	ahead := &ListNode{0, head}
	for prev, cur := ahead, head ;cur != nil; cur = cur.Next {
		if cur.Val == val {
			prev.Next = cur.Next
		} else {
			prev.Next = cur
			prev = cur
		}
	}
	return ahead.Next
}
// @lc code=end

